### DISCUSSION TITLE
Can Oclif auto-load plugins that are installed alongside the CLI

### DISCUSSION THREAD

**Discussion by moltar** [oclif/oclif#1235] [Q&A]
Would it be possible to do something like this:

```sh
npm i -D @org/cli @org/plugin-foo
```

And then `@org/cli` would be magically aware of `@org/plugin-foo` existence and load it?

That is without explicitly specifying the plugin in `oclif.plugins` list.

I am also aware of `@oclif/plugin-plugins`, but it's not what I want, but similar. I want to manage plugins exclusively via npm.

**Comment by mdonnalley** [comment]
It's certainly possible but I question whether it's actually practical since oclif would need to traverse the entire node_modules to find packages that resemble a plugin. You could cache the result but that first traversal would be (I'm assuming) very slow, especially on windows machines.

You might be able to accomplish this using npm's lifecycle scripts. For example, you might be able to use the `postinstall` script to add the newly installed plugins to the `oclif.plugins` list - or something like that.

**Reply by moltar** [reply]
> newly installed plugins to the oclif.plugins

Are you suggesting to modify the `package.json` of the installed CLI package?

**Reply by mdonnalley** [reply]
Yes, that's what I had in mind. To be clear, that was just the first idea off the top of my head - I have no idea whether or not it's feasible

**Reply by moltar** [reply]
I think it's a bad idea, especially in cases of pnpm, which keeps packages in a central store, and then links them in `node_modules`.

**Comment by mdonnalley** [comment]
@moltar would defining plugins using wildcards accomplish what you want? I'm thinking oclif could use `minimatch` to find the matching plugins from `dependencies`

```json
{
  "dependencies": {
    "@oclif/core": "^3",
    "@oclif/plugin-help": "^6",
    "@oclif/plugin-plugins": "^4",
    "@org/cool-plugin": "^1",
    "@org/awesome-plugin": "^1",
   },
  "oclif": {
    "plugins": [
        "@oclif/plugin-*",
        "@org/*",
     ]
  }
}
```

**Reply by moltar** [reply]
Ya I think that's a great idea! That way one can allowlist org plugins only and avoid rogue plugins getting thru.

**Comment by mdonnalley** [comment]
This is now supported in @oclif/core@3.22.0

**Comment by davidensinger** [comment]
@mdonnalley sorry to bring this back from the dead, pluginsAdditions works great, except that I’d like to allow for plugins to be denoted as "devDependencies" in the consuming project instead of "dependencies", but I'm not sure I want the other behavior that comes along with setting `development: true` in execute() (or setting process.env.NODE_ENV to "development"). Has there been any consideration to either:

- Allow for "devDependencies" in loadCorePlugins in plugin-loader.ts
- Remove the isProd (or workaround somehow?) guard in loadDevPlugins in plugin-loader.ts

The reason I’d prefer to denote plugins as "devDependencies" is that they’d be developer tooling/CLIs and not client side/production "dependencies" that would run in a browser.

**Reply by mdonnalley** [reply]
I'm having trouble understanding your use case. A CLI can still have client side dependencies even though it doesn't run in a web browser. And if you intend to publish your CLI on npm (or anywhere else), the plugins won't be installed when the user installs your CLI since they're listed as devDependencies

**Reply by davidensinger** [reply]
I’d like to have a root/canonical CLI, say `@myorg/cli`. I’d like to configure it to support loading plugins that are peers, say `@myorg/cli-plugin-*`, via the pluginsAdditions configuration. Each little extensible plugin would have optional functionality that not everyone would necessarily want/need, which is why I wouldn’t want to put it by default in the root/canonical CLI.

This is what a consuming package.json would typically look like for my use case:

```json
  "dependencies": {
    "react": "~18.3.1",
    "react-dom": "~18.3.1"
    ...other client side dependencies...
  },
  "devDependencies": {
    "@myorg/cli": "~1.0.0",
    "@myorg/cli-plugin-someplugin": "~1.0.0",
    "@myorg/cli-plugin-anotherplugin": "~1.0.0",
    "jest": "~29.7.0,
    "msw": "~2.3.1",
    "typescript": "~5.5.2"
    ...other developer tooling dependencies...
  }
```

If I denote the plugin as a "core" plugin with this configuration:

```ts
pluginAdditions = {
  core: ['@myorg/cli-plugin-*'],
  path: process.cwd(),
}
```

It won’t find the plugin because it’s in not in "dependencies" https://github.com/oclif/core/blob/main/src/config/plugin-loader.ts#L98.

If I denote the plugin as "dev" plugin:

```ts
pluginAdditions = {
  dev: ['@myorg/cli-plugin-*'],
  path: process.cwd(),
}
```

It will only find the plugin if I also set `devPlugins = true` and set `process.env.NODE_ENV = 'development'` or set `development: true` when I init with execute() (to get past this https://github.com/oclif/core/blob/main/src/config/plugin-loader.ts#L108-L110). Denoting these plugins as devPlugins works for me, I'm just not sure of the other ramifications of setting NODE_ENV to "development" for consumers of the CLI and plugin(s). Since they would not be doing development work on the CLI or plugins themselves, they’d only be using them locally as part of their day-to-day dev workflow. So for my use case, I’d prefer if core plugins for pluginAdditions would also look in "devDependencies" https://github.com/oclif/core/blob/main/src/config/plugin-loader.ts#L98 because I do not want to denote any of these plugins as "dependencies" in the canonical CLI since they’d be completely optional, so I wouldn’t want to install them for everyone.

Thank you for your quick reply, I do really appreciate it, as I know this is often thankless work.

**Reply by mdonnalley** [reply]
> I’d like to have a root/canonical CLI, say @myorg/cli. I’d like to configure it to support loading plugins that are peers, say @myorg/cli-plugin-*, via the pluginsAdditions configuration. 

Correct me if I'm wrong: it sounds like you want users to be able to independently add plugins to the CLI? Why not use https://github.com/oclif/plugin-plugins then? That way users can run `plugins install @myorg/cli-plugin-someplugin` if they decide that that's a plugin they want to use

>Denoting these plugins as devPlugins works for me, I'm just not sure of the other ramifications of setting NODE_ENV to "development" for consumers of the CLI and plugin(s)

Off the top of my head, the only ramifications are errors and warnings will show entire stack traces. Also, performances could be considerably slower

**Reply by davidensinger** [reply]
Yeah, I played with plugins-plugins but I didn't like:

- that a user would need to be prompted to install a plugin (or multiple plugins) versus a narrow allowlist that searches in node_modules for plugins/commands. Like if it’s already in node_modules it would just work.
- that the installed user plugins live in that local dataDir instead of node_modules. I found this functionality to be surprising. I'm also a little unsure what would happen for an end user who had different versions of a plugin installed in different projects. Would this support multiple versions?

These are minor quibbles, though, so again thank you for hearing me out. This is for my work and we’ve rolled our own CLI functionality for so long, it seems like oclif will be a big improvement for us.
